From 7367300774200a3779875dba33d0717627c175bc Mon Sep 17 00:00:00 2001
From: Ben Horgan <ben.horgan@arm.com>
Date: Tue, 26 Apr 2022 18:46:14 +0000
Subject: [PATCH 22/22] spmc S-EL2: Get the cpu num.

A read of MPIDR_EL1 at EL1 returns the value of VMPIDR_EL2

Signed-off-by: Ben Horgan <ben.horgan@arm.com>
Change-Id: I23a45ac917d9094d15acd2402abe9e7625a82025
Upstream-Status: Pending [Not submitted to upstream yet]
---
 platform/generic-arm64/plat_asm.S | 32 +++++++++++++++++++++++++++++++
 platform/generic-arm64/rules.mk   |  3 +++
 2 files changed, 35 insertions(+)
 create mode 100644 platform/generic-arm64/plat_asm.S

diff --git a/platform/generic-arm64/plat_asm.S b/platform/generic-arm64/plat_asm.S
new file mode 100644
index 0000000..e736b3c
--- /dev/null
+++ b/platform/generic-arm64/plat_asm.S
@@ -0,0 +1,32 @@
+/*
+ * Copyright (c) 2022, Arm Limited. All rights reserved.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining
+ * a copy of this software and associated documentation files
+ * (the "Software"), to deal in the Software without restriction,
+ * including without limitation the rights to use, copy, modify, merge,
+ * publish, distribute, sublicense, and/or sell copies of the Software,
+ * and to permit persons to whom the Software is furnished to do so,
+ * subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be
+ * included in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+ * IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+ * CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+ * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+ * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+#include <asm.h>
+#include <arch/asm_macros.h>
+
+#if TRUSTY_ARM_ARCH_SECURE_EL2
+FUNCTION(arm64_curr_cpu_num)
+    mrs     x0, mpidr_el1
+    ubfx    x0, x0, #0, #8
+    ret
+#endif
diff --git a/platform/generic-arm64/rules.mk b/platform/generic-arm64/rules.mk
index 1c04bfe..9671a45 100644
--- a/platform/generic-arm64/rules.mk
+++ b/platform/generic-arm64/rules.mk
@@ -75,4 +75,7 @@ MODULE_INCLUDES += \
 MODULE_SRCS += \
 	$(LOCAL_DIR)/smc_service_access_policy.c \
 
+MODULE_SRCS += \
+	$(LOCAL_DIR)/plat_asm.S \
+
 include make/module.mk
-- 
2.17.1

